client/engine/messageservice/p2p-message-service/service.go

```diff
@@ -0,0 +1,194 @@
+package p2pms
+
+import (
+	"bufio"
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"math/rand"
+	"time"
+
+	"github.com/libp2p/go-libp2p"
+	p2pcrypto "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/statechannels/go-nitro/client/engine/store/safesync"
+	"github.com/statechannels/go-nitro/crypto"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+const (
+	MESSAGE_ADDRESS      = "/messages/1.0.0"
+	DELIMITER            = '\n'
+	BUFFER_SIZE          = 1_000
+	NUM_CONNECT_ATTEMPTS = 20
+	RETRY_SLEEP_DURATION = 5 * time.Second
+)
+
+// P2PMessageService is a rudimentary message service that uses TCP to send and receive messages
+type P2PMessageService struct {
+	out chan protocols.Message // for sending message to engine
+
+	peers *safesync.Map[peer.ID]
+
+	quit    chan struct{} // quit is used to signal the goroutine to stop
+	me      types.Address
+	key     p2pcrypto.PrivKey
+	p2pHost host.Host
+}
+
+// Id returns the libp2p peer ID of the message service
+func (ms *P2PMessageService) Id() peer.ID {
+	id, _ := peer.IDFromPrivateKey(ms.key)
+	return id
+}
+
+// AddPeers adds the peers to the message service
+// We ignore peers that are ourselves
+func (ms *P2PMessageService) AddPeers(peers []PeerInfo) {
+
+	for _, p := range peers {
+		// Ignore ourselves
+		if p.Address == ms.me {
+			continue
+		}
+		multi, _ := multiaddr.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d/p2p/%s", p.IpAddress, p.Port, p.Id))
+
+		// Extract the peer ID from the multiaddr.
+		info, _ := peer.AddrInfoFromP2pAddr(multi)
+		// Add the destination's peer multiaddress in the peerstore.
+		// This will be used during connection and stream creation by libp2p.
+		ms.p2pHost.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.PermanentAddrTTL)
+
+		ms.peers.Store(p.Address.String(), info.ID)
+	}
+}
+
+// NewMessageService returns a running P2PMessageService listening on the given ip and port
+func NewMessageService(ip string, port int, pk []byte) *P2PMessageService {
+	// We generate a random key using the hash of the pk as a seed
+	// This should mean that the message key is deterministic
+	// TODO: Ideally we would use the pk directly, but I haven't figured out of this is possible with lib p2p
+	hash := sha256.Sum256(pk)
+	seed := big.NewInt(0).SetBytes(hash[:]).Int64()
+	messageKey, _, err := p2pcrypto.GenerateECDSAKeyPair(rand.New(rand.NewSource(seed)))
```

Comment:
I believe the key for lib p2p uses a different curve than ours, and isn't directly compatible but I'm not really sure.